recipe/meta.yaml

```diff
@@ -51,10 +56,10 @@ outputs:
         - if not exist %LIBRARY_PREFIX%\\bin\\dnnl.dll exit 1  # [win]
         - if not exist %LIBRARY_PREFIX%\\lib\\dnnl.lib exit 1  # [win]
         - if not exist %LIBRARY_PREFIX%\\include\\dnnl.h exit 1  # [win]
-  - name: {{ name|lower }}-cpu-{{ dnnl_cpu_runtime }}
+  - name: {{ name|lower }}-cpu-{{ dnnl_cpu_runtime }}-gpu-{{ dnnl_gpu_runtime }}
```

Comment:
Wouldn't it make sense to have these separate?
For eg:
```
onednn-cpu-tbb onednn-gpu-dpcpp
```

Reply:
oneDNN has ability support for both CPU and GPU devices simultaneously, so a user can build a library that uses TBB threading for CPU functionality and DPCPP runtime for GPU functionality, that is why I mixed them together, so there are multiple possible configs:
- onednn-cpu-tbb-gpu-dpcpp
- onednn-cpu-none-gpu-dpcpp
- onednn-cpu-dpcpp-gpu-dpcpp
- onednn-cpu-tbb-gpu-none

Is this the corect way of implementing it in conda?

Reply:
Yes, but this is about how users would request those packages. A user should be able to do
```
conda install onednn-gpu-dpcpp
```
if they only care about the GPU and does not have to specify which CPU variant to use.

With the current PR, the user has to specify both CPU and GPU variants.

Reply:
Yes, this is the intention since different CPU variants require different runtimes and will impact the user's application. Usually each user chooses the relevant CPU. For example, OpenVino could request `onednn-gpu-dpcpp` that does not have a CPU runtime at all, because they use separate plugins for CPU and GPU. IPEX (Intel Extentions for PyTorch) would use `onednn-gpu-dpcpp` but with OpenMP for CPU and TensorFlow might want to have `onednn-gpu-dpcpp` but with Threadpool for CPU because their execution model uses Eigen threadpool. It would be good to have some aliases like `onednn-cpu-tbb` for `onednn-cpu-tbb-gpu-none` and similar things for OpenMP and Threadpool, but I don't know yet how to do it.

Reply:
If user tries to install 2 different onednn packages, i would expect `conda install onednn-cpu-tbb onednn-gpu-dpcpp` will install `onednn-cpu-tbb-gpu-none` and `onednn-cpu-none-gpu-dpcpp` which are valid and supported oneDNN configurations. If a user wants a single package that works with both CPU and GPU then `conda install onednn-cpu-tbb-gpu-dpcpp` should work for them.

Reply:
Here is an example: https://anaconda.org/intel/onednn-cpu-dpcpp-gpu-dpcpp 
I think we could minimize amount of configurations to the existing ones in main branch + cpu-dpcpp-gpu-dpcpp, how does it sound to you?

Reply:
> If user tries to install 2 different onednn packages

We don't allow that. It has to be only one onednn package. I see that `onednn-cpu-dpcpp-gpu-dpcpp` and `onednn-cpu-gomp` in intel channel clobber each other since both of them have `libdnnl.so.*` and override each other.

Reply:
O, I see, I didn't know about this issue. How does it work if a user needs multiple configurations of a library?

Reply:
> How does it work if a user needs multiple configurations of a library within one conda environment?

You would have to rename the libraries. `libdnnl.so` -> `libdnnl-cpu-gomp.so`.

This is not special for conda environment. It's a packaging/installation issue in general.

Reply:
Yeah, I didn't mean it is specific to conda, I didn't realize conda has it as well. I will discuss with the team what is the best way to proceed. Thank you a lot for the review, it is very helpful!

Reply:
BTW, would it be possible to forbid installing 2 packages by the user to have all configurations available but allow only one of them to be installed in a given environment?

Reply:
> BTW, would it be possible to forbid installing 2 packages by the user to have all configurations available but allow only one of them to be installed in a given environment?

Yes, it's done already since there is one `onednn` package with variants that `onednn-cpu-*-gpu-*` is depending on.

Reply:
Ok, great, I was worried overrriding happens silently